datafusion/physical-expr/src/array_expressions.rs

```diff
@@ -433,7 +433,7 @@ pub fn array_element(args: &[ArrayRef]) -> Result<ArrayRef> {
             let indexes = as_int64_array(&args[1])?;
             general_array_element::<i64>(array, indexes)
         }
-        _ => exec_err!(
+        _ => not_impl_err!(
```

Comment:
Most array functions only accept `List`, `LargeList`, and `FixedSizeList`. So it should not be  `not_impl_err`.

Reply:
So ... in your opinion what should the error be? All I can think of what I had previously (exec_err), what @comphead suggested (not_impl_err) or a brand new error (say invalid_type_err)? 

Essentially every single one of the changes I made was for this pattern, arrays are not special in this regard.

Reply:
Because most of them are related to unsupported type errors, a new brand error `invalid_type_err` could be more accurate and readable.

Reply:
I think exec_err is a good choice. THe value of a new structured error is that some program could react specially to it, but I can't think of a good reason that the reaction to a "invalid_type_err" would be different than a exec_err. Thus I don't think a new type is a good idea in this case

Reply:
I reverted the not_impl_err commit, it just doesn't feel correct to me.